tests/unit/s2n_early_data_test.c

```diff
@@ -0,0 +1,162 @@
+
+        /* Test known valid / invalid transitions */
+        {
+            struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+            EXPECT_NOT_NULL(conn);
+
+            conn->early_data_state = 0;
+            EXPECT_ERROR_WITH_ERRNO(s2n_connection_set_early_data_state(conn, S2N_UNKNOWN_EARLY_DATA_STATE),
+                    S2N_ERR_INVALID_EARLY_DATA_STATE);
+
+            conn->early_data_state = 0;
+            EXPECT_ERROR_WITH_ERRNO(s2n_connection_set_early_data_state(conn, S2N_EARLY_DATA_STATES_COUNT),
+                    S2N_ERR_INVALID_EARLY_DATA_STATE);
+
+            conn->early_data_state = 0;
+            EXPECT_ERROR_WITH_ERRNO(s2n_connection_set_early_data_state(conn, S2N_EARLY_DATA_STATES_COUNT + 1),
+                    S2N_ERR_INVALID_EARLY_DATA_STATE);
+
+            conn->early_data_state = 0;
+            EXPECT_OK(s2n_connection_set_early_data_state(conn, S2N_EARLY_DATA_REQUESTED));
+
+            conn->early_data_state = S2N_EARLY_DATA_REQUESTED;
+            EXPECT_OK(s2n_connection_set_early_data_state(conn, S2N_EARLY_DATA_REJECTED));
+
+            conn->early_data_state = S2N_EARLY_DATA_REQUESTED;
+            EXPECT_ERROR_WITH_ERRNO(s2n_connection_set_early_data_state(conn, S2N_END_OF_EARLY_DATA),
+                    S2N_ERR_INVALID_EARLY_DATA_STATE);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+        }
+
+        /* Test that only the expected sequences of states are possible.
+         * Given every possible sequence of early data states, test that s2n_connection_set_early_data_state
+         * can only be used to iterate through the known valid sequences. */
+        {
+            /* Test with the correct expected sequences */
+            {
+                const s2n_early_state_sequence valid_early_state_sequences[] = {
+                        { .states = early_data_not_requested_seq, .len = s2n_array_len(early_data_not_requested_seq) },
+                        { .states = early_data_rejected_seq, .len = s2n_array_len(early_data_rejected_seq) },
+                        { .states = early_data_success_seq, .len = s2n_array_len(early_data_success_seq) },
+                };
+
+                struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+                EXPECT_NOT_NULL(conn);
+                EXPECT_OK(s2n_test_all_early_data_sequences(conn, 0,
+                        valid_early_state_sequences, s2n_array_len(valid_early_state_sequences)));
+                EXPECT_SUCCESS(s2n_connection_free(conn));
+            }
+
+            /* Sanity check: adding an invalid expected sequence causes test to fail */
+            {
+                const s2n_early_data_state invalid_seq[] = {
+                        S2N_UNKNOWN_EARLY_DATA_STATE, S2N_EARLY_DATA_ACCEPTED, S2N_END_OF_EARLY_DATA };
+                const s2n_early_state_sequence valid_early_state_sequences[] = {
```

Comment:
Nit: technically this valid_early_state_sequences array contains an invalid sequence.